src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Interlocked.Wasm.cs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime;
+using System.Runtime.CompilerServices;
+using System.Diagnostics.CodeAnalysis;
+
+using Internal.Runtime.CompilerHelpers;
+
+namespace System.Threading
+{
+    //
+    // WASM atomics trap on unaligned access so we must manually check the alignment of locations before
+    // calling the underlying instructions, to throw a managed misalignment exception. Likewise with null
+    // reference exception. We delegate this to codegen where possible, so most of methods below are
+    // always-expand intrinsics.
+    //
+    // TODO-LLVM: once/if upstream factoring of atomics changes, reconsider the need for a separate file.
+    //
+    public static partial class Interlocked
+    {
+        [Intrinsic]
+        public static unsafe int CompareExchange(ref int location1, int value, int comparand) => CompareExchange(ref location1, value, comparand);
+
+        [Intrinsic]
+        public static unsafe long CompareExchange(ref long location1, long value, long comparand) => CompareExchange(ref location1, value, comparand);
+
+        [Intrinsic]
+        [return: NotNullIfNotNull(nameof(location1))]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static T CompareExchange<T>(ref T location1, T value, T comparand) where T : class?
+        {
+            return Unsafe.As<T>(CompareExchange(ref Unsafe.As<T, object?>(ref location1), value, comparand));
+        }
+
+        [Intrinsic]
+        [return: NotNullIfNotNull(nameof(location1))]
+        public static unsafe object? CompareExchange(ref object? location1, object? value, object? comparand)
+        {
+            if (Unsafe.IsNullRef(ref location1))
+            {
+                ThrowHelpers.ThrowNullReferenceException();
+            }
+            if (((nint)Unsafe.AsPointer(ref location1) & (sizeof(nint) - 1)) != 0)
```

Comment:
`ref object` is always aligned. The type system and the runtime should guarantee that. I think this can be assert - it does not need to be a runtime check.